source/extensions/quic_listeners/quiche/platform/envoy_quic_clock.cc

```diff
@@ -4,13 +4,15 @@ namespace Envoy {
 namespace Quic {
 
 quic::QuicTime EnvoyQuicClock::ApproximateNow() const {
-  // This might be expensive as Dispatcher doesn't store approximate time_point.
-  return Now();
+  return quic::QuicTime::Zero() + quic::QuicTime::Delta::FromMicroseconds(microsecondsSinceEpoch(
+                                      dispatcher_.approximateMonotonicTime()));
 }
 
 quic::QuicTime EnvoyQuicClock::Now() const {
-  return quic::QuicTime::Zero() + quic::QuicTime::Delta::FromMicroseconds(microsecondsSinceEpoch(
-                                      dispatcher_.timeSource().monotonicTime()));
+  // Since the expensive operation of obtaining time has to be performed anyway,
+  // make Dispatcher update approximate time.
+  const_cast<Event::Dispatcher&>(dispatcher_).updateApproximateMonotonicTime();
```

Comment:
The const_cast here is a bit of a code smell. Can we hold a reference to a non-const dispatcher?

Reply:
Doesn't smell good to me either.  Unfortunately EnvoyQuicClock::Now() method is const, therefore it needs to work when all members are const.  EnvoyQuicClock::dispatcher_ is a non-const reference BTW.

Either quic::QuicClock::Now() that we override here needs to be changed to non-const in QUICHE, or we leave this const_cast here so that the code compiles.  Or we can muck around with a privately owned class instance decorated to be mutable that wraps the dispatcher reference.

One could of course argue that if calling Now() updates ApproximateNow(), then Now is not really const anyway.

Cf. three commits ago when EnvoyQuicClock owned and updated an approximate_time_ member, that member was marked mutable for the same reason.

Reply:
OK this seems problematic but if you want to add a TODO here and come back to it later that's ok. Up to you.

Reply:
I spent some time investigating what it would entail to change QuicClock::Now() to non-const in QUICHE.  I don't think it's worth it, not because it's a lot of work, but because it adds confusion by forcing many other methods and member to become non-const.  For example, QuicClock::ConvertWallTimeToQuicTime(), QuicConnection::clock(), and QuicConnectionHelperInterface::GetClock().  Also, QuicPacketReader::ReadAndDispatch*Packet*() cannot take QuicClock by const ref.  This is a can of worms not worth opening.

I would argue that QuicClock::Now() is still conceptually const, and QuicClock::ApproximateNow() should not have any guarantees about when exactly it is updated.  In fact I was second guessing myself while writing tests for it being updated on Now() calls and on event loops that this tests more the implementation than the interface.

As far as this code smell, I think const_cast is exactly as bad as a mutable member, so I don't believe this can be made any better.

I propose that we land this PR without a TODO.  I can add a comment to platform/api/quic_clock.h in QUICHE about guarantees of ApproximateNow() if you think it's helpful.  WDYT?

Reply:
Honestly I think we are prematurely optimizing and I would just remove the update within the `Now()` call, but if you don't want to do that can you at least add a comment that is a summary of this conversation explaining the const_cast?

Reply:
I would prefer keeping the update within Now() because of the way how Envoy timer works. ApproximateNow() is used in QuicAlarm impl to subtract Now() or ApproximateNow() to get the duration value which is used to schedule timer. Not updating approximate time when Now() is called may lead to alarms being scheduled slightly later than desired. Although I don't know how much this could affect QUICHE, but having it updated can avoid such issue.

Reply:
After talking to Dan we concluded that it is possible that without this optimization alarms fire one event loop later.  For example, EnvoyQuicAlarmTest.PostponeDeadline fails with the alarm not firing.

I drafted the commit f8e66766262719d84c7b46f3165b843eed206565 to not update approximate time from Now, and indeed it's simpler, but not much simpler.  Sure this optimization is a little premature, but the cost is mostly code complexity.  The performance hit is minimal.  And since timeSource().monotonicTime() is somewhat costly, it feels silly not to update approximate time every time it's called.

On the other hand, I know that an alarm firing one event loop later is not a huge deal.

In any case, I'd like to keep the logic that updates approximate time from Now().  I'll add the comment you suggested.

Reply:
That's fine, sounds good.